Implemented reality checks for Inverse Hyperbolic Functions

HarryMountain opened the pull request:
Implemented _eval_is_real for inverse hyperbolic functions: asinh, acosh, atanh and acoth

Fixes #24928 
Fixes #25612

<!-- BEGIN RELEASE NOTES -->
* functions
  * Implemented reality checks for Inverse Hyperbolic functions
<!-- END RELEASE NOTES -->

HarryMountain committed:
Inverse Hyperbolic: Implemented reality checks

Fixed issue #24928.

HarryMountain committed:
Added name to mailmap

oscarbenjamin commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -1826,6 +1835,9 @@ def inverse(self, argindex=1):
         return coth
 
+    def _eval_is_real(self):
+        return self.args[0].is_real and not -1 < self.args[0] < 1
```
```
In [5]: x = symbols('x', real=True)

In [6]: atanh(x).is_real
...
TypeError: cannot determine truth value of Relational
```

oscarbenjamin commented on sympy/functions/elementary/tests/test_hyperbolic.py:
```diff
@@ -678,6 +681,11 @@ def test_acosh():
+    # Reality
+    assert acosh(2).is_real
+    assert not acosh(0.5).is_real
+    assert (acosh(2) - oo) == -oo
```
It is better to assert that True, False etc are actually returned explicitly:
```python
assert acosh(2).is_real is True
```
We don't want `is_real` to return any other "truthy" object. Same goes for asserting with `not` which would pass for either False or None.

Note that `acosh` with a Float evaluates:
```
In [7]: acosh(0.5)
Out[7]: 1.0471975511966⋅ⅈ
```
It would be better to test with a Rational:
```python
In [10]: acosh(S(1)/5)
Out[10]: acosh(1/5)

In [11]: acosh(S(1)/5).is_real
Out[11]: False
```

HarryMountain committed:
Updated reality checks on inverse hyperbolic functions

Handling Symbolic arguements.
More accurate tests.

HarryMountain commented:
Thank you very much for your feedback, I have addressed all of your comments however while fixing them I found another bug that I raised Issue #25612 for.

oscarbenjamin commented:
Actually it would be better to implement `_eval_is_extended_real` and `_eval_is_finite` rather than `_eval_is_real`. Their combination can resolve an `is_real` query but having the two handlers separately means that we can also resolve queries like `is_extended_real` and `is_finite` better as well.

oscarbenjamin commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -1540,6 +1543,9 @@ def _eval_is_zero(self):
             return True
 
+    def _eval_is_real(self):
+        return self.args[0].is_real and (self.args[0] - 1).is_positive
```
Potentially checking `is_real` here is redundant because the `is_positive` query would be False if `is_real` is False. I prefer not to create the expression `arg - 1` though so it is useful to have that shortcut.

HarryMountain committed:
Implemented _eval_is_extended_real and _eval_is_finite instead of _eval_is_real on inverse hyperbolic functions.

HarryMountain committed:
Implemented _eval_is_extended_real and _eval_is_finite on asech and acsch to facilitate reality checks.

oscarbenjamin commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -1543,8 +1546,11 @@ def _eval_is_zero(self):
-    def _eval_is_real(self):
-        return self.args[0].is_real and (self.args[0] - 1).is_positive
+    def _eval_is_extended_real(self):
+        return self.args[0].is_extended_real and (self.args[0] - 1).is_positive
```
Should this be `(self.args[0] - 1).is_extended_positive`?

oscarbenjamin commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -1689,8 +1695,18 @@ def _eval_is_zero(self):
+        if self.args[0].is_extended_real:
+            return (1 - self.args[0]).is_positive and (self.args[0] + 1).is_positive
+
+    def _eval_is_finite(self):
+        check_against_one = (self.args[0] - 1).is_zero
+        if check_against_one is None:
+            return None
+        check_against_minus_one = (self.args[0] + 1).is_zero
+        if check_against_minus_one is None:
+            return None
+        return check_against_one is False and check_against_minus_one is False
```
Is this different from
```
return fuzzy_not(fuzzy_or((self.args[0] - 1).is_zero, (self.args[0] + 1).is_zero))
```

oscarbenjamin commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -2020,6 +2046,19 @@ def _eval_rewrite_as_atanh(self, x, **kwargs):
+            return check_against_zero is True and check_against_one is True
+
+    def _eval_is_finite(self):
+        return self.args[0].is_zero is not True
```
If `is_zero` gives None should this return True?

oscarbenjamin commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -2213,3 +2252,13 @@ def _eval_rewrite_as_atanh(self, arg, **kwargs):
+
+    def _eval_is_finite(self):
+        is_zero = self.args[0].is_zero
+        return None if is_zero is None else not is_zero
+
+
+
```
There are some blank lines added here at the end of the file.

oscarbenjamin commented:
Three-valued logic is really hard!

HarryMountain committed:
Improved inverse hyperbolic functions reality checks to use fuzzy logic

HarryMountain commented:
Thank you very much for your feedback. It is extremely helpful and I am learning a lot.

oscarbenjamin commented:
Looks good. Thanks!

gschintgen commented:
Hi @HarryMountain, thanks for this PR. I was wondering why some rather trivial intersections with `S.Reals` created internally by `solveset` were simplifying nicely in my development branch, while they would not on 1.12. Good work.
